compiler-lib/src/type_checking/builtin_types.rs

```diff
@@ -0,0 +1,95 @@
+use crate::strtab::StringTable;
+
+use super::{checker::SemanticContext, type_system::*};
+
+pub struct BuiltinTypes<'src> {
+    pub string: CheckedType<'src>,
+}
+
+impl<'src> BuiltinTypes<'src> {
+    pub fn add_to<'ts>(
+        type_system: &'ts mut TypeSystem<'src>,
+        strtab: &'_ mut StringTable<'src>,
+        context: &'_ mut SemanticContext<'_, 'src>,
+    ) -> BuiltinTypes<'src> {
+        let int_ty = CheckedType::Int;
+
+        let (_, reader_class_id) = {
```

Comment:
These `let (_, reader_class_id) = { /* function body */ };` are really weird. You should probably add a comment explaining why its written that way.

Reply:
Could you add a few markdown paragraphs to the type_checking/mod.rs that explain implementation details like the special meaning of "$"?

Reply:
will take care of that